### ISSUE TITLE
processing references with `chapterbib.sty` and `splitat`

### ISSUE THREAD

**Issue by evoludolab** [brucemiller/LaTeXML#1206]
Some referencing issues exist when using `chapterbib.sty` together with the `--splitat=chapter` option. The following minimal (but unfortunately rather complex) example reproduces the issue.

<details> 
<summary>Consider the four files `chapterbib.tex`, `chapterbib1.tex`, `chapterbib2.tex` and `chapterbib.bib` (show details to view contents or download archive below).</summary>

```
*** chapterbib.tex: ***
\documentclass{book}
\usepackage{chapterbib}
% produce sections with references
\usepackage[sectionbib]{natbib}
\renewcommand{\bibname}{References}

\begin{document}
\include{chapterbib1}
\include{chapterbib2}

\cleardoublepage
\renewcommand{\bibname}{Bibliography}
% produce chapter with bibliography
\renewcommand\bibsection{\chapter*{\bibname}\addcontentsline{toc}{chapter}{\bibname}}

% generate overall bibliography
\bibliography{chapterbib}

\end{document}

*** chapterbib1.tex: ***
\chapter{\label{ch:one}One}

some references: \cite{ref1} and \cite{ref2}

\bibliographystyle{abbrvnat}
\bibliography{chapterbib}

*** chapterbib2.tex: ***
\chapter{\label{ch:two}Two}

some more references: \cite{ref1} (same as in \ref{ch:one}) and \cite{ref3}

\bibliographystyle{abbrvnat}
\bibliography{chapterbib}

*** chapterbib.bib: ***
@article{ref1,
	Author = {Author 1},
	Journal = {Journal 1},
	Pages = {1-10},
	Title = {Title 1},
	Volume = {1},
	Year = {2001}}

@article{ref2,
	Author = {Author 2},
	Journal = {Journal 2},
	Pages = {2-20},
	Title = {Title 2},
	Volume = {2},
	Year = {2002}}

@article{ref3,
	Author = {Author 3},
	Journal = {Journal 3},
	Pages = {3-30},
	Title = {Title 3},
	Volume = {3},
	Year = {2003}}

@article{ref4,
	Author = {Author 4},
	Journal = {Journal 4},
	Pages = {4-40},
	Title = {Title 4},
	Volume = {4},
	Year = {2004}}
```

</details>

Using `latex` the following sequence of commands produces the desired output:
```
pdflatex chapterbib.tex
bibtex chapterbib
bibtex chapterbib1
bibtex chapterbib2
pdflatex chapterbib.tex
pdflatex chapterbib.tex
```
with a reference section at the end of each chapter and another chapter with the overall bibliography at the end of the document (see enclosed chapterbib.pdf).

For `LaTeXML` the command sequence:
```
pdflatex chapterbib.tex
latexml --dest=chapterbib.xml chapterbib.tex
latexmlpost --prescan --dbfile=chapterbib.db --splitat=chapter --dest=chapterbib.xhtml chapterbib.xml
latexmlpost --noscan --dbfile=chapterbib.db --splitat=chapter --dest=chapterbib.xhtml chapterbib.xml
```
gets close. However, this may simply be an incorrect usage of `latexml` on my part. More specifically, the second call to `latexmlpost` reports
```
(CrossRef Ch1.xhtml processing...
Info:expected:ids Missing items:
	  Usable title for ID: TEMPORARY_DOCUMENT_ID
	[Note TEMPORARY_DOCUMENT_ID is a stand-in ID for the main document.]
 0.00 sec)
(CrossRef Ch2.xhtml processing...
Warning:expected:ids Missing items:
	  Entry for citation: ref3
	In Post::CrossRef[@0x7fcb66720580] ->process
...
```
with the effect that the references at the end of chapter 1 and 2 appear as expected but the citation/link to ref3 in chapter 2 cannot be resolved (although the citation to ref1, which already appeared in chapter 1 is resolved, see enclosed Ch2.xhtml). Moreover, the overall bibliography at the end is empty (see enclosed bibb.xhtml).

Some further related comments/suggestions:
1. the title of the reference section is formatted as a chapter, i.e. has the same style classes as the bibliography at the end and hence makes it difficult to adjust/customize styling using CSS.
2. each citation includes a convenient 'Cited by:' list with references to the citing text. For bibliographies at the end of a chapter it would be nice if those references were restricted to the corresponding chapter (while referencing all appearances in the overall bibliography at the end).

For convenience the whole set of files is enclosed.
[chapterbib.zip](https://github.com/brucemiller/LaTeXML/files/3613246/chapterbib.zip)

**Comment by evoludolab** [comment]
After looking into the issue a bit more I think I managed to narrow things down a bit. 

Without `--splitat` everything works as expected. However, when adding `--documentid=myid` it seems that `Post::CrossRef.pm` messes up the automatically generated labels for links and anchors but the strange naming remains inconsequential without `--splitat`.

More specifically, the different bibliographies are labelled `bib`, `biba`, `bibb` etc. and result in links/anchors to entries in the corresponding bibliography labelled `bib.bib1`, `biba.bib1`, etc. When specifying a document id, the bibliographies are labelled `myid.bib`, `myid.biba`, `myid.bibb`, etc. as expected. However, the links/anchors become `bibmyid.bib.bib1` and `bibamyid.bib.bib1` etc. instead of `myid.bib.bib1` and `myid.biba.bib1` (note the duplication of `bib` and the placement of the `a`).

While I suspect the required fix to be simple (and hopefully resolves the problems with `splitat` as well), `CrossRef.pm` turns out to be too challenging for me to navigate for now...

**Comment by dginev** [comment]
Thanks for the extra diagnostic @evoludolab ! This is reminiscent of a clash I also had with `--documentid` back in the day, so if I was to look into the issue I would vet the full id interplay with respect to that option. Maybe Bruce can see this more clearly since he's the cross-referencing guru, but I can take a look soon as well.

**Comment by evoludolab** [comment]
Couldn't resist some more digging... Turns out the problem is not with `CrossRef.pm` but more subtle:

When processing the `bib` file, `latexml` produces a `bib.xml` file with `bibentry`'s that have `xml:id=myid.bib.bib1` etc. This results in the labelling issues when those entries are used to build different bibliographies.

It seems to me that the `bib.xml` file should only use `xml:id=bib1` etc. while the document id (here `myid`) and bibliography id (`bib`, `biba`, etc.) should only be added when processing the actual references. Or, more precisely, both id's are correctly added but this happens in addition to the `xml:id` already present in `bib.xml`.

As an aside, such a change may have consequences when using multiple `bib`-files that I cannot anticipate (and won't need).

**Comment by brucemiller** [comment]
Thanks for the debugging; that pointed me exactly to the right part of the program, where I found a nice comment about "isn't ideal for multiple bibliographies". Not "ideal"!?!?! What optimism!

The fix will take a bit more digging...

**Comment by brucemiller** [comment]
Sorry for the delay, the need for "more diggiing" allowed me to get completely sidetracked.  As I commented on your PR: it got me pointed in the right direction, but didn't deal with the "lists" infrastructure that was meant to deal with these kinds of issues, but hadn't quite gotten finished as it applied to bibliographies.  I *think* I've got it right, and it seems to handle all your issues (except for the distinct class on bibliography sections). See 9ebd6633.  Thanks for the report & PR!!
